### PR TITLE
Support listening to multiple events at the same time

### LINKED ISSUE #1: Support multiple events in `on()`
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/emittery/issues/1)

I considered adding support for multiple events in `on()`, but wanted to wait to see if this is something people actually need.

Relevant Node.js issue: https://github.com/nodejs/node/issues/15697

// @silverwind

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#36 Support multiple events in on, off and once](https://issuehunt.io/r/sindresorhus/emittery/pull/36)
- [#52 support multiple events in on, off and once](https://issuehunt.io/r/sindresorhus/emittery/pull/52)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/emittery/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by LuKks**
Fix #1
This pull request is based on the one by @Angelmmiguel #36

Closes #36

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#1: Support multiple events in `on()`](https://issuehunt.io/repos/112248999/issues/1)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -87,15 +103,34 @@ Only events that are not of this type are able to trigger these events.
...
+	await emitter.emit('🦄', 'a');
+	await emitter.emit('🐶', 'b');
+	await emitter.emit('🦊', 'c');
+	emitter.off('🦄', listener);
+	emitter.off(['🐶', '🦊'], listener);
+	await emitter.emit('🦄', 'a'); // nothing happens
+	await emitter.emit('🐶', 'b'); // nothing happens
+	await emitter.emit('🦊', 'c'); // nothing happens
+})();
+```
```
Comment (sindresorhus) [inline readme.md:127]: The docs in the readme and index.d.ts should be in sync. That means all examples too.

File: index.d.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ Emittery accepts strings and symbols as event names.
 Symbol event names can be used to avoid name collisions when your classes are extended, especially for internal events.
 */
 type EventName = string | symbol;
+type MultiEventName = string | string[] | symbol;
```
Comment (sindresorhus) [inline index.d.ts:7]:
```suggestion
type EventNames = EventName | readonly EventName[];
```

Comment (LuKks) [reply index.d.ts]: Ooh, I really liked that one.

File: index.d.ts [hunk 3]
```diff
@@ -4,6 +4,7 @@ Emittery accepts strings and symbols as event names.
 Symbol event names can be used to avoid name collisions when your classes are extended, especially for internal events.
 */
 type EventName = string | symbol;
+type EventNames = EventName | readonly EventName[];
```
Comment (sindresorhus) [inline index.d.ts:7]: This needs a doc comment too.

File: test/index.js [hunk 4]
```diff
@@ -201,6 +212,24 @@ test('off()', async t => {
 	t.deepEqual(calls, [1]);
 });
 
+test('off() - multiple event names', async t => {
+	const emitter = new Emittery();
+	const calls = [];
+	const listener = () => calls.push(1);
```
Comment (sindresorhus) [inline test/index.js:218]:
```suggestion
	const listener = () => {
		calls.push(1);
	};
```

Don't use inline arrow function if it doesn't actually return anything. Applies in many other places too.

File: test/index.js [hunk 5]
```diff
@@ -14,6 +14,17 @@ test('on()', async t => {
 	t.deepEqual(calls, [1, 2]);
 });
```
Comment (sindresorhus) [inline test/index.js:16]: I'd like to see tests for symbols and and string names being mixed.

File: readme.md [hunk 6]
```diff
@@ -108,8 +143,12 @@ emitter.once('🦄').then(data => {
 	console.log(data);
 	//=> '🌈'
 });
+emitter.once(['🦄', '🐶']).then(data => {
+	console.log(data);
+});
 
-emitter.emit('🦄', '🌈');
+emitter.emit('🦄', '🌈'); // log => '🌈' x2
+emitter.emit('🐶', '🍖');	// nothing happens
```
Comment (sindresorhus) [inline readme.md:151]:
```suggestion
emitter.emit('🐶', '🍖'); // nothing happens
```

File: index.js [hunk 7]
```diff
@@ -187,33 +187,44 @@ class Emittery {
...
 
-	once(eventName) {
+	once(eventNames) {
 		return new Promise(resolve => {
-			assertEventName(eventName);
-			const off = this.on(eventName, data => {
+			const events = Array.isArray(eventNames) ? eventNames : [eventNames];
+			for (const eventName of events) {
+				assertEventName(eventName);
+			}
```
Comment (sindresorhus) [inline index.js:225]: `.on` already asserts the event names.

File: index.js [hunk 8]
```diff
@@ -187,33 +187,44 @@ class Emittery {
...
 
-	on(eventName, listener) {
-		assertEventName(eventName);
+	on(eventNames, listener) {
 		assertListener(listener);
-		getListeners(this, eventName).add(listener);
 
-		if (!isListenerSymbol(eventName)) {
-			this.emit(listenerAdded, {eventName, listener});
+		const events = Array.isArray(eventNames) ? eventNames : [eventNames];
```
Comment (sindresorhus) [inline index.js:193]:
```suggestion
		eventNames = Array.isArray(eventNames) ? eventNames : [eventNames];
```

File: test/index.js (hunk 4 above)
Comment (LuKks) [reply test/index.js]: I also changed it in other places that are not related to this PR, I hope that is ok.

**Comment by sindresorhus** [comment]
I just noticed that the supports is not complete. You also need to support multiple names in:

- https://github.com/LuKks/emittery/blob/fe4e9b4793f5e8018c652a98afe8e9d6149ccd1e/index.d.ts#L165
- https://github.com/LuKks/emittery/blob/fe4e9b4793f5e8018c652a98afe8e9d6149ccd1e/index.d.ts#L288

**Comment by sindresorhus** [comment]
Thanks :)

**Pull request merged**